crates/subspace-runtime/src/lib.rs

```diff
@@ -478,25 +487,112 @@ impl pallet_rewards::Config for Runtime {
+    type BlockNumber = u32;
+    type Hash = <BlakeTwo256 as Hasher>::Out;
+    type Header = generic::Header<u32, BlakeTwo256>;
+}
+
 /// Type used to represent a FeedId or ChainId
 pub type FeedId = u64;
+pub struct GrandpaValidator<C>(C);
 
-pub struct GrandpaValidator;
-
-impl FeedValidator<FeedId> for GrandpaValidator {
-    fn initialize(feed_id: FeedId, data: &[u8]) -> DispatchResult {
+impl<C: Chain> FeedProcessor<FeedId> for GrandpaValidator<C> {
+    fn init(&self, feed_id: FeedId, data: &[u8]) -> sp_runtime::DispatchResult {
         pallet_grandpa_finality_verifier::initialize::<Runtime>(feed_id, data)
     }
 
-    fn validate(feed_id: FeedId, object: &[u8]) -> DispatchResult {
-        pallet_grandpa_finality_verifier::validate::<Runtime>(feed_id, object)
+    fn put(&self, feed_id: FeedId, object: &[u8]) -> Result<Option<FeedMetadata>, DispatchError> {
+        Ok(Some(
+            pallet_grandpa_finality_verifier::validate_finalized_block::<Runtime, C>(
+                feed_id, object,
+            )?
+            .encode(),
+        ))
+    }
+
+    fn object_mappings(&self, _feed_id: FeedId, object: &[u8]) -> Vec<FeedObjectMapping> {
+        let block = match C::decode_block::<Runtime>(object) {
+            Ok(block) => block,
+            // we just return empty if we failed to decode as this is not called in runtime
+            Err(_) => return vec![],
+        };
+        // for substrate, we store the block as is. so offset is 0
+        vec![FeedObjectMapping {
+            key: block.block.header.hash().into(),
+            offset: 0,
+        }]
+    }
+
+    fn delete(&self, feed_id: FeedId) -> sp_runtime::DispatchResult {
+        pallet_grandpa_finality_verifier::purge::<Runtime>(feed_id)
+    }
+}
+
+pub struct ParachainImporter<C>(C);
+
+impl<C: Chain> ParachainImporter<C> {
+    fn decode_block(
+        &self,
+        object: &[u8],
+    ) -> Result<pallet_grandpa_finality_verifier::chain::SignedBlock<C::Header>, DispatchError>
+    {
+        let block = C::decode_block::<Runtime>(object)?;
+        Ok(block)
+    }
+}
+
+impl<C: Chain> FeedProcessor<FeedId> for ParachainImporter<C> {
+    fn put(&self, _feed_id: FeedId, object: &[u8]) -> Result<Option<FeedMetadata>, DispatchError> {
+        let block = self.decode_block(object)?;
+        Ok(Some(
+            (block.block.header.hash(), *block.block.header.number()).encode(),
+        ))
+    }
+    fn object_mappings(&self, _feed_id: FeedId, object: &[u8]) -> Vec<FeedObjectMapping> {
+        self.decode_block(object)
+            .ok()
+            .map(|block| {
+                vec![FeedObjectMapping {
+                    key: block.block.header.hash().into(),
+                    offset: 0,
+                }]
+            })
+            .unwrap_or_default()
+    }
+}
+
+/// FeedProcessorId represents the available FeedProcessor impls
+#[derive(Debug, Clone, Copy, Encode, Decode, TypeInfo, Eq, PartialEq)]
+pub enum FeedProcessorId {
```

Comment:
Something like `FeedProcessorKind` or `FeedProcessorType` could be clearer for me.